User folder selection + lots of other stuff

CST1229 opened the pull request:
This PR was originally planned to just add a user selection to fix compiling executables, but it ended up becoming... this.

Resolves #18
Partially addresses #16 (adds a user folder path, but you still have to specify the runtimes folder too)
Resolves #15
Resolves #13
Resolves #5

This is my first time doing really anything related to GMEdit plugins, so there might be lots of bad code and issues.

- You can now specify a GameMaker user folder (%AppData%/GameMakerStudio2) and user in the control panel. This fixes compiling executables and allows using the IDE config and license files for YYC and <=2022 builds.
- Compiling on Windows and Mac now uses the PackageZip verb instead of trying (and failing) to use Package.
- Global runtime options actually save now.
- LTS and 2022 builds are now supported.
- Added support for YYC builds. I call the VM/YYC choice the "runner", to distinguish it from the Stable/Beta/LTS or runtime version choices that are called "runtime". 
- Igor is now passed the `/v` (verbose) flag, which GM also uses.
- Job tabs now display different finished messages for when builds fail or are cancelled.
- The runtime version can now be selected per-project.
- "Use Default" runtime choices are now displayed correctly in the control panel.

CST1229 committed:
lots of stuff

CST1229 committed:
Automatically set unset users/runtimes
also fix some == comparisons

CST1229 committed:
remove an old leftover and tweak igor paths a bit

thennothinghappened commented:
Hello!

Firstly just want to say thank you so much for this - I've not had time to do this work on Constructor so I _really_ appreciate this. I had a look over it this morning briefly and the main issues I have are nitpicks just to make sure there's a consistent style going on. I'll add those as review comments when I can but overall looks good. I'll give this a test on my machine when I'm in GMEdit.

thennothinghappened commented on js/ui/editors/CompileLogViewer.js:
```diff
@@ -28,7 +28,9 @@ class KConstructorOutput extends ConstructorViewFileKind {
     static getJobName = (job) => {
-        return `${job.projectDisplayName} - ${job.command}${job.stopped ? ' - Finished' : ''}`;
+        return `${job.projectDisplayName} - ${job.command}${
+            job.stopped ? (' - ' + (job.exitCode === -1 ? 'Stopped' : ((job.exitCode !== null && job.exitCode > 0) ? 'Failed' : 'Finished'))) : ''
```
This one really needs to become some actual log - drawing the line at nested ternaries lol

CST1229 replied:
i didn't turn it into an actual log but i cleaned up the nested ternaries by making it into a property on the Job object

thennothinghappened commented on js/compiler/igor-controller.js:
```diff
@@ -13,16 +12,24 @@ import { spawn } from '../GMConstructor.js';
 /** @type {Job[]} */
 const jobs = [];
 
+const fs = require("node:fs/promises");
```
I know it's a little strange but check out the `init.js` script - in essence since GMEdit plugins aren't in ES6 land properly and don't get to use imports I used some silly workarounds to get that working with a dynamic import. Part of the reason was it also meant the TS language server was happy. Side effect is that it isn't happy with trying to import node things via the `import` syntax, so my workaround was piping them in via the `init` script.

Last I remember mixing `require`s in to do this made the LSP very upset and I'm not a fan of mixing them myself either so would prefer this stick with the current setup - see `GMConstructor.js` for what that looks like too.

thennothinghappened commented on js/compiler/igor-controller.js:
```diff
@@ -13,16 +12,24 @@ import { spawn } from '../GMConstructor.js';
+
+    // for some reason if we don't clear the cache directory, changes won't apply in yyc
+    if (settings.runner === "YYC") {
+        await fs.rm(project.dir + "/cache/", {recursive: true, force: true});
```
I'm not super comfortable with using what's basically an `rm -rf` here, I'd rather leave that to Igor to manage which it already does via a `Clean` task. If this is needed I'd argue for invoking a `Clean` and on completion invoking the actual compile job. I'd want to do more experimenting on this too since the cache is a big part of YYC speed and initial compilation on YYC is horrendously slow.

thennothinghappened commented on js/compiler/igor-controller.js:
```diff
@@ -13,16 +12,24 @@ import { spawn } from '../GMConstructor.js';
+export async function job_run(project, runtime, user, settings) {
+
+    // for some reason if we don't clear the cache directory, changes won't apply in yyc
+    if (settings.runner === "YYC") {
```
For consistency make sure to use single quotes in this project :P

thennothinghappened commented on js/compiler/igor-controller.js:
```diff
@@ -72,20 +79,43 @@ function job_remove(job) {
+
+    let projectName = $gmedit['gml.Project'].current.name;
+    // Remove .yyp file extension
+    projectName = projectName.substring(0, projectName.length - 4);
```
I'd prefer steering away from removing extensions and stuff with magic numbers, I _think_ node's fs module has something to do this for you? We've already got that imported for other path stuff so would be better to apply it here.

CST1229 replied:
solved by just doing `project.displayName` instead of `project.name` and removing the .yyp file extension

thennothinghappened commented on js/compiler/igor-controller.js:
```diff
@@ -72,20 +79,43 @@ function job_remove(job) {
-        `/runtime=${settings.runtime}`
+        `/runtime=${settings.runner}`,
+        `/v`,
+        `/tf=${project.dir}/output/${projectName}${exts[igor_platform_cmd_name] ?? ""}`
```
GMEdit exposes a `displayName` and `name` if you query the project, see here:
https://github.com/thennothinghappened/GMEdit-Constructor/blob/main/js/global.d.ts#L369

thennothinghappened commented on js/compiler/igor-controller.js:
```diff
@@ -72,20 +79,43 @@ function job_remove(job) {
+        Linux: ".appimage",
+    };
+
+    let projectName = $gmedit['gml.Project'].current.name;
```
I'd prefer this uses `project_current_get()` from utils rather than querying directly - genuinely just a nitpick that i think looks nicer and clearer.

thennothinghappened commented on js/compiler/igor-controller.js:
```diff
@@ -72,20 +79,43 @@ function job_remove(job) {
     switch (settings.verb) {
-        case 'Run':
         case 'Package':
+            if (['Windows', 'Mac'].includes(igor_platform_cmd_name))
```
This is a style thing and applies for all single line `if`s - I avoid them in all of my projects so here I am enforcing that too, apologies if that wasn't clear, I wasn't really expecting anyone to contribute so have no style guide specified. The `{}` are mandatory!!

thennothinghappened commented on js/compiler/igor-paths.js:
```diff
@@ -2,6 +2,9 @@
 import { join_path } from '../GMConstructor.js';
 import { Err } from '../utils/Err.js';
 
+const process = require('node:process');
```
Second instance of `require`, also I think we already have `process` somewhere, or did earlier...

CST1229 replied:
I saw the init.js script before, and considered it while doing this for igor-paths, but decided to use `require` here because this exports `igor_platform_map` directly, instead of being in a function or something like that, and I can't really pass arguments into it that way. VSCode doesn't complain to me aout mixing `import` and `require` (and yes it does about type-related things so it is working), though I'm not sure how you use the TS language server.
I'll see what I could do.

thennothinghappened commented on js/compiler/job/Job.js:
```diff
@@ -18,6 +18,9 @@ export class Job {
     #stopped = false;
+    
+    /** @type {number|null} */
+    #exitCode = null;
```
I'd prefer going with something like a union type for something like this, where if we need the exit code it could be something more clear like:

```ts
declare type JobStatus = 
    { status: 'running' } |
    { status: 'stopped', stoppedByUser: boolean, exitCode: number };
```

And then `#stopped` and `#exitCode` can be wrapped into one while providing some nicer intellisense, and with no sentinel `-1` value :P

thennothinghappened commented on js/global.d.ts:
```diff
@@ -98,6 +112,14 @@ declare type RuntimeInfo = {
+/**
+ * Information for a specific found user.
+ */
+declare type UserInfo = {
```
The next exciting thing from this is it opens up reading `devices.json`, so happy to see this!

thennothinghappened commented on js/preferences/Preferences.js:
```diff
@@ -103,6 +126,33 @@ export function save_on_run_task_set(save_on_run_task) {
+ * Get the desired runner type.
+ * @returns {RunnerType}
+ */
+export function runner_get() {
```
I think the naming convention wasn't overly clear in retrospect so later on I might change these all (ie the similar ones I did and your new ones) to use `prop_get_or_fallback()` or something. More verbose but better distinction than `prop_get()` vs `prop_project_get()` since they're both querying the project properties.

thennothinghappened commented on js/preferences/Preferences.js:
```diff
@@ -201,6 +286,53 @@ export async function runtime_search_path_set(type, search_path) {
+        const err = new Err(
+            `Failed to load ${type} user list`,
+            res.err,
+            'Make sure the users path is valid!'
```
I appreciate leaving in the "how can you fix this" lines :P

CST1229 replied:
I really just copy-pasted the runtimes code for most of the user folder-related code.

thennothinghappened commented on js/preferences/Preferences.js:
```diff
@@ -379,14 +562,20 @@ export async function __setup__() {
-    prefs = Object.create(prefs_default);
-    Object.assign(prefs, loaded_prefs);
+    prefs = structuredClone(prefs_default);
+    deep_assign(prefs, loaded_prefs);
```
I'm a bit curious why foregoing `Object.create` for `structuredClone` here, the goal was to fallback to the default values when none are specified.

CST1229 replied:
I did this because `JSON.stringify` does not stringify the prototypes of objects, like what happens when doing `Object.create`. This is fine for "top-level" properties, since those get assigned and copied over into the "actual" properties, but for properties inside other objects (which is the case for runtime options in the config file), the object in the prototype gets mutated instead and JSON.stringify passes over it, causing it to not be saved into the config file.
![image](https://github.com/thennothinghappened/GMEdit-Constructor/assets/68464103/f689f24d-bae8-440f-bdd7-0f7672dd878c)

thennothinghappened commented on js/preferences/Preferences.js:
```diff
@@ -406,6 +595,38 @@ export async function __setup__() {
+            options.user = users[runtime_type][0].name.toString();
+        }
+    }
+
```
This one when I did it was already on the verge of needing to be switched out for a nicer `for...of [Stable, Beta, LTS]...` type thing, I think at this point it definitely needs that lol.

thennothinghappened commented on js/compiler/igor-controller.js:
```diff
@@ -72,20 +79,43 @@ function job_remove(job) {
+        Windows: ".zip",
+        Mac: ".zip",
+        Linux: ".appimage",
+    };
```
I'd prefer to move this into `igor-paths.js` with the other platform-dependent filename things.

thennothinghappened commented on js/preferences/Preferences.js:
```diff
@@ -103,6 +126,33 @@ export function save_on_run_task_set(save_on_run_task) {
+ * The default runner type used globally.
+ * @returns {RunnerType}
+ */
+export function runner_project_get() {
```
I'm not really sure why there's two of these

CST1229 replied:
I think I did the runner getter/setter functions for ProjectProperties first, then pasted and modified into Preferences but forgot to remove the `_project` functions that aren't needed for it

thennothinghappened commented on js/ui/HamburgerOptions.js:
```diff
@@ -188,10 +188,10 @@ export class HamburgerOptions {
      * @param {boolean} enabled
      */
     #setEnableMenuItems = (enabled) => {
-        this.#menu_items_container.enabled = enabled;
```
Good catch, that was a leftover from before the control panel

thennothinghappened commented on js/ui/PreferencesMenu.js:
```diff
@@ -33,8 +33,9 @@ export function __cleanup__() {
+ * @param {Function?} [on_change_runtime_channel]
  */
-export function menu_create(prefs_group) {
+export function menu_create(prefs_group, on_change_runtime_channel) {
```
I'm not a fan of how this is now passing in a callback, especially given the callback isn't documented in the JSDoc either. I was using a lot of callbacks (see much earlier commits on this) but it got very... unenjoyable lol. I'm not sure if anything I immediately think of would replace this better, just not a massive fan of this. Should be fine though.

thennothinghappened replied:
I think this kinda makes me want events for preferences updates that the control panel can listen for, but that's not something relevant to this PR and is something I'll do later if I've got time.

thennothinghappened commented on js/ui/editors/ConstructorControlPanel.js:
```diff
@@ -319,20 +324,71 @@ export class ConstructorControlPanel extends ConstructorEditorView {
+        UIPreferences.addDropdown(
+            this.projectSettings,
+            'Runner Type',
+            projectProperties.runner_project_get() ?? 'Use Default',
```
Not a fan of how much of a sentinel value `'Use Default'` is getting, I'd prefer to store it somewhere I think just so there's no possible typing error :P

thennothinghappened commented on js/ui/editors/ConstructorControlPanel.js:
```diff
@@ -344,8 +400,27 @@ export class ConstructorControlPanel extends ConstructorEditorView {
-        preferencesMenu.menu_create(this.globalSettings);
-
+        preferencesMenu.menu_create(this.globalSettings, () => {
+            if (!this.#runtime_version_dropdown)
```
Another nitpick - please make sure to use explicit comparisons over implicit JS `!` for checking not null.

thennothinghappened commented on js/utils/object.js:
```diff
@@ -0,0 +1,24 @@
+ * @param {any} obj
+ * @returns {boolean}
+ */
+export function is_object(obj) {
```
This doesn't seem overly necessary, and is also exported despite being used once - in the function directly below.

thennothinghappened commented on js/utils/object.js:
```diff
@@ -0,0 +1,24 @@
+ */
+export function deep_assign(to, obj) {
+	if (!is_object(obj) || !is_object(to)) return to;
+	for (const key of Object.keys(obj)) {
```
I'm not a fan of this because it treats object-like things as proper objects, e.g. arrays:

```js
console.log([] instanceof Object); // true
console.log(Object.keys(['hi'])); // ['0']
```

That does _work_, but it feels icky.

thennothinghappened commented on js/ui/editors/ConstructorControlPanel.js:
```diff
@@ -319,20 +324,71 @@ export class ConstructorControlPanel extends ConstructorEditorView {
             [...preferences.valid_runtime_types, 'Use Default'],
             (value) => {
 
                 if (value === 'Use Default') {
```
When the callbacks are getting this long I'd prefer they're proper callbacks on the class itself

thennothinghappened reviewed (changes requested):
Overall great work, I'd really like you to stick to the style though and clean up the `include` usage among some of the other things I've mentioned. From there I'll give this a go on my laptop and then should be good to merge. Thanks!

CST1229 committed:
address review comments
- clean up uses of `require` outside init.js
- move exts to from igor-controller igor-paths (can't put it inside igor_platform_map because that uses node.js platform names and exts is target-dependent, not host-os-dependent)
- rename appdata variable to windowsAppdata
- code style stuff
- make `job.#status` into an object, replacing `job.#exitCode`
- make the status display a variable on the job object instead of a big nested ternary
- fix building with a Use Default project runtime channel not working

CST1229 committed:
single quotes

CST1229 committed:
another implicit null check

thennothinghappened commented:
Looks like you've fixed up most of what I pointed at, thanks again!!
Again I'll have to look at it properly tomorrow at the soonest though, its quite late and I'm quite sick at the moment.

thennothinghappened reviewed (commented):
Mainly looking at the `rm` usage, otherwise looks pretty good!

CST1229 committed:
review comments 2
- use the project argument in job_flags_get
- pass /ic to igor in yyc to ignore the cache instead of deleting the whole cache folder (this probably still leaves compiling slow but it at least doesn't delete a whole folder)
- clean up if statements in preferences runtime loading

CST1229 commented on js/compiler/igor-controller.js:
```diff
@@ -17,12 +17,18 @@ const jobs = [];
+
+    // for some reason if we don't clear the cache directory, changes won't apply in yyc
+    if (settings.runner === 'YYC') {
+        await rm(project.dir + '/cache/', {recursive: true, force: true});
```
apparently passing `/ic` (ignorecache) to Igor instead works, though idk if it still causes a speed loss or not

thennothinghappened reviewed (commented):
Thanks for being so persistent with my feedback :)
I'll give this a run on my machine and see how it goes!

thennothinghappened commented:
Also glad to see `rm` swapped out for flags, I'm much happier with that.

thennothinghappened commented:
Looks like Node isn't a big fan of paths using `~`:

<img width="549" alt="image" src="https://github.com/thennothinghappened/GMEdit-Constructor/assets/35243139/e5bd9092-3199-4f47-bf30-ab4f0cf3eec6">

I'd just swap `~` for `process.env.HOME`.

thennothinghappened commented:
<img width="579" alt="image" src="https://github.com/thennothinghappened/GMEdit-Constructor/assets/35243139/0396d9af-ad5c-4060-a6dd-7264568f9c8d">

I've done that here and it fixed it on my end.

thennothinghappened commented:
<img width="532" alt="image" src="https://github.com/thennothinghappened/GMEdit-Constructor/assets/35243139/3b1a4c66-c023-4640-991e-1df75e58f757">

Not entirely sure why but we seem to have lost the GMEdit "normal" preferences.
The control panel is the "proper" way for this now so I might have a look at fixing this as well later to get this merged sooner.

thennothinghappened commented:
Can confirm YYC working on MacOS, fantastic!!

<img width="1114" alt="image" src="https://github.com/thennothinghappened/GMEdit-Constructor/assets/35243139/6b5df68d-22b6-497f-a5ea-e83c2ac4d48c">
